crates/ruma-state-res/src/event_auth/room_member.rs

```diff
@@ -0,0 +1,588 @@
+    let target_user_level =
+        power_levels.users.get(target_user).unwrap_or(&power_levels.users_default);
+
+    // If the sender’s power level is greater than or equal to the ban level, and the
+    // target user’s power level is less than the sender’s power level, allow.
+    //
+    // Otherwise, reject.
+    let allow = sender_level >= &power_levels.ban && target_user_level < sender_level;
+
+    if !allow {
+        warn!(
+            sender_membership_event_id = ?sender_room_member_event.as_ref().map(|event| event.event_id()),
+            power_levels_event_id = ?room_power_levels_event.as_ref().map(|event| event.event_id()),
+            "sender does not have enough power to ban target user",
+        );
+    }
+
+    Ok(allow)
+}
+
+/// Check whether the given event passes the `m.room.member` authorization rules with a membership
+/// of `knock`.
+fn check_room_member_knock<E: Event>(
+    room_member_event: &impl Event,
+    target_user: &UserId,
+    room_version: &RoomVersion,
+    fetch_state: impl Fn(&StateEventType, &str) -> Option<E>,
+) -> Result<bool> {
+    let room_join_rules_event = fetch_state(&StateEventType::RoomJoinRules, "");
+    let join_rule = match &room_join_rules_event {
+        Some(event) => from_json_str::<RoomJoinRulesEventContent>(event.content().get())?.join_rule,
+        None => JoinRule::Invite,
+    };
+
+    // v7-v9, if the join_rule is anything other than knock, reject.
+    // Since v10, if the join_rule is anything other than knock or knock_restricted,
+    // reject.
+    if join_rule != JoinRule::Knock
+        && (room_version.knock_restricted_join_rule
+            && !matches!(join_rule, JoinRule::KnockRestricted(_)))
+    {
+        warn!("join rule is not set to knock or knock_restricted, knocking is not allowed");
+        return Ok(false);
+    }
+
+    // Since v7, if sender does not match state_key, reject.
+    if room_member_event.sender() != target_user {
+        warn!(
+            sender = ?room_member_event.sender(),
+            ?target_user,
+            "can't make another user knock, sender did not match target"
+        );
+        return Ok(false);
+    }
+
+    let sender_room_member_event =
+        fetch_state(&StateEventType::RoomMember, room_member_event.sender().as_str());
+    let sender_membership = match &sender_room_member_event {
+        Some(event) => from_json_str::<GetMembership>(event.content().get())?.membership,
+        None => MembershipState::Leave,
+    };
+
+    // Since v7, if the sender’s current membership is not ban, invite, or join, allow.
+    // Otherwise, reject.
+    let allow = !matches!(
+        sender_membership,
+        MembershipState::Ban | MembershipState::Invite | MembershipState::Join
+    );
+
+    if !allow {
+        warn!(
+            target_user_membership_event_id = ?sender_room_member_event.as_ref().map(|event| event.event_id()),
+            "membership state of ban, invite or join are invalid",
```

Comment:
This is a bit ambiguous, as the membership states themselves are valid, just not the transition
```suggestion
            "cannot knock while current membership state is ban, invite or join",
```

Reply:
Applied